# Statistics 
" Statistics is the discipline of collecting, analyzing, interpreting, and presenting data to identify patterns and inform decision-making. "

In [5]:
# lets say we have marks of students so how will we draw points from this 
marks  =  [45, 56, 67, 78, 89, 90, 56, 67, 78, 56]

from the above list we can have many questions like : 
1. Average marks ?
2. Common marks in my list ?
3. Middle student score ?

In [22]:
#imports
import numpy as np
import pandas as pd

## Mean 
Mean is defined as the Sum of all 'n' observations / Total number of observations 

$$
\text{Mean} = \frac{\sum x}{n}
$$

Where:

- $\sum x$ = Sum of all observations
- $n$ = Number of observations

In [10]:
marks = np.array([10,20,30,40,50])
mean = np.mean(marks)
print(mean)

30.0


for this case the mean is totally reliable and useful in getting the avg but for the same scenario if we update the array to have one outliers or just one single large value the avg suddendly spikes irrelevant with the data

In [13]:
marks = np.array([10,20,30,40,1000])
print(np.mean(marks))

220.0


Earlier the mean suggested that If everyone's marks were distributed equally, each student would get 30 marks.
And now , the question arises that did a student even score 220 ? No, One extreme value (1000) distorted the average which is called as an Outlier

### Outlier
By definition, it is an extreme endpoint (a value far away from the rest of the data).
By effect, it is the value that distorts the average (the mean).

## Median
Median is the middle observation in a sorted dataset.
For odd number of observations:
$$
Median = Middle\ Value
$$
For even number of observations:
$$
Median =
\frac{
Middle_1 + Middle_2
}{2}
$$

In [17]:
eg = [10,20,30,35,40,50]
print(np.median(marks))

30.0


In [18]:
# In case of an outlier
eg = [10,20,30,35,40,1000]
print(np.median(marks))

30.0


Median didn't get affected. This is why salaries are often reported using median instead of mean.

## Mode

Mode is the observation with the highest frequency.

$$
Mode = Most\ Frequent\ Value
$$

In [23]:
marks = pd.Series([10,20,20,20,30,40])
print(marks.mode())

0    20
dtype: int64


Mode tells us : The most common value. Useful in:
1. Product sizes
2. Most purchased item
3. Most common age group

Summary of the day : 
| Measure | Meaning | Affected by Outliers? |
|----------|----------|----------------------|
| Mean | Average | Yes |
| Median | Middle Value | No |
| Mode | Most Frequent Value | No |

In [30]:
data = {
    "Marks":[45, 56, 67, 78, 89, 90, 56, 67, 78, 56,500]
}

df = pd.DataFrame(data)

print("Mean:",df["Marks"].mean())
print("Median:",df["Marks"].median())
print("Mode:",df["Marks"].mode()[0])

Mean: 107.45454545454545
Median: 67.0
Mode: 56


> When should you use Median instead of Mean?

When data contains outliers or is skewed. Median is more robust because extreme values do not affect it significantly.